source/common/protobuf/utility.cc

```diff
@@ -27,6 +27,48 @@ using namespace std::chrono_literals;
 namespace Envoy {
 namespace {
 
+absl::Status validateDurationNoThrow(const ProtobufWkt::Duration& duration,
+                                     int64_t max_seconds_value) {
+  if (duration.seconds() < 0 || duration.nanos() < 0) {
+    return absl::OutOfRangeError(
+        fmt::format("Expected positive duration: {}", duration.DebugString()));
+  }
+  if (duration.nanos() > 999999999 || duration.seconds() > max_seconds_value) {
+    return absl::OutOfRangeError(fmt::format("Duration out-of-range: {}", duration.DebugString()));
+  }
+  return absl::OkStatus();
+}
+
+void validateDuration(const ProtobufWkt::Duration& duration, int64_t max_seconds_value) {
+  const auto result = validateDurationNoThrow(duration, max_seconds_value);
+  if (!result.ok()) {
+    throwEnvoyExceptionOrPanic(std::string(result.message()));
+  }
+}
+
+void validateDuration(const ProtobufWkt::Duration& duration) {
+  validateDuration(duration, Protobuf::util::TimeUtil::kDurationMaxSeconds);
+}
+
+void validateDurationAsMilliseconds(const ProtobufWkt::Duration& duration) {
```

Comment:
Is this validating as nanoseconds or milliseconds? Not clear why we check overflow for nanos but not millis.

Reply:
I think that the actual issue is that while protobof converts both seconds and milliseconds to nanoseconds, our validation treats them differently.
I'll work on aligning both to the same limit.